src/rust/lib/graph-editor/src/component/visualization/registry.rs

```diff
@@ -0,0 +1,97 @@
+//! a specific data type.
+//!
+//! Example
+//! --------
+//! ```no_run
+//! use graph_editor::component::visualization::Registry;
+//! use graph_editor::component::visualization::EnsoType;
+//! use graph_editor::component::visualization::JsSourceClass;
+//!
+//! // Instantiate a pre-populated registry.
+//! let registry = Registry::with_default_visualizations();
+//! // Add a new class that creates visualizations defined in JS.
+//! registry.register_class(JsSourceClass::from_js_source_raw(r#"
+//!     class BubbleVisualization {
+//!         static inputTypes = ["[[float;3]]"]
+//!         onDataReceived(root, data) {}
+//!         setSize(root, size) {}
+//!     }
+//!     return BubbleVisualization;
+//! "#.into()).unwrap());
+//!
+//! // Get all factories that can render  visualization for the type `[[float;3]]`.
+//! let target_type:EnsoType = "[[float;3]]".to_string().into();
+//! assert!(registry.valid_sources(&target_type).len() > 0);
+//! ```
+
+use crate::prelude::*;
+
+use crate::component::visualization::*;
+use crate::component::visualization::renderer::example::js::get_bubble_vis_class;
+use crate::component::visualization::renderer::example::native::BubbleChart;
+
+use ensogl::display::scene::Scene;
+
+
+
+// ==============================
+// === Visualization Registry ===
+// ==============================
+
+/// HashMap that contains the mapping from `EnsoType`s to a `Vec` of `Factories. This is meant to
+/// map a `EnsoType` to all `visualization::Class`es that support visualising that type.
+type RegistryTypeMap = HashMap<EnsoType, Vec<Rc<dyn Class>>>;
+
+/// The registry struct. For more information see the module description.
+#[derive(Clone,CloneRef,Default,Debug)]
+#[allow(missing_docs)]
+pub struct Registry {
+    entries : Rc<RefCell<RegistryTypeMap>>,
+}
+
+impl Registry {
+    /// Return an empty `Registry`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Return a `Registry` prepopulated with default visualizations.
+    pub fn with_default_visualizations() -> Self {
+        let registry = Self::new();
+        // FIXME use proper enso types here.
+        registry.register_class(NativeConstructorClass::new(
+            ClassAttributes {
+                name        : "Bubble Visualization (native)".to_string(),
+                input_types : vec!["[[float;3]]".to_string().into()],
+            },
+            |scene:&Scene| Ok(Visualization::new(BubbleChart::new(scene)))
+        ));
+        registry.register_class(get_bubble_vis_class());
+
+        registry
+    }
+
+    /// Register a new visualization class with the registry.
+    pub fn register_class<T: Class + 'static>(&self, class:T) {
+        self.register_class_rc(Rc::new(class));
+    }
+
+    /// Register a new visualization class that's pre-wrapped in an `Rc` with the registry.
+    pub fn register_class_rc(&self, class:Rc<dyn Class>) {
```

Comment:
Please, do not expose rc'd dyn traits in public API. Create newtype for them, always. Or make this function private.